agent/src/protocols/coap_core/discovery_handler.rs

```diff
@@ -0,0 +1,416 @@
+impl DiscoveryHandler for CoAPCoREDiscoveryHandler {
+    async fn discover(&self) -> Result<Vec<DiscoveryResult>, anyhow::Error> {
+        let multicast = self.discovery_handler_config.multicast;
+        let static_addrs = &self.discovery_handler_config.static_ip_addresses;
+        let multicast_addr = &self.discovery_handler_config.multicast_ip_address;
+        let query_filter = self.discovery_handler_config.query_filter.as_ref();
+        let timeout =
+            Duration::from_secs(self.discovery_handler_config.discovery_timeout_seconds as u64);
+        let mut results: Vec<DiscoveryResult> = vec![];
+
+        // Discover devices via static IPs
+        for ip_address in static_addrs {
+            let coap_client = CoAPClientImpl::new((ip_address.as_str(), 5683));
+            let result = self.discover_endpoint(&coap_client, ip_address, query_filter, timeout);
+
+            match result {
+                Ok(result) => results.push(result),
+                Err(e) => {
+                    info!("Discovering endpoint {} went wrong: {}", ip_address, e);
+                }
+            }
+        }
+
+        // Discover devices via multicast
+        if multicast {
+            let coap_client = CoAPClientImpl::new((multicast_addr.as_str(), 5683));
+            let discovered = self.discover_multicast(&coap_client, query_filter, timeout);
+
+            match discovered {
+                Ok(mut rs) => {
+                    results.append(&mut rs);
+                }
+                Err(e) => {
+                    return Err(anyhow::format_err!(
+                        "Error while discovering devices via multicast {}",
+                        e
+                    ));
+                }
+            }
+        }
+
+        Ok(results)
+    }
+
+    fn are_shared(&self) -> Result<bool, anyhow::Error> {
+        Ok(false)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::net::{IpAddr, Ipv4Addr};
+
+    use super::*;
+    use akri_shared::coap_core::MockCoAPClient;
+    use coap_lite::{CoapResponse, MessageType, Packet};
+    use mockall::predicate::eq;
+
+    fn create_core_response() -> CoapResponse {
+        let mut request = Packet::new();
+        request.header.set_type(MessageType::Confirmable);
+
+        let mut response = CoapResponse::new(&request).unwrap();
+
+        response.message.payload = br#"</sensors/temp>;rt="oic.r.temperature";if="sensor",
+                </sensors/light>;rt="oic.r.light.brightness";if="sensor""#
+            .to_vec();
+
+        response
+    }
+
+    fn configure_unicast_response(mock: &mut MockCoAPClient, timeout: Duration) {
+        mock.expect_get_with_timeout()
+            .withf(move |_url, tm| *tm == timeout)
+            .returning(|_url, _timeout| Ok(create_core_response()));
+    }
+
+    #[tokio::test]
+    async fn test_discover_resources_via_ip_addresses() {
+        // TODO: find better way than setting env variable, which could be shared to other tests
```

Comment:
The issue lies in https://github.com/deislabs/akri/blob/1c81b6ca44cccb596f0b19a01a220d96154b7658/agent/src/protocols/mod.rs#L24

which needs to read the env variable. It could receive a trait as parameter, but it feels a bit hacky because it should be passed down all the way from the tested method so that we can pass a mock. I'm open tu suggestions.

Reply:
Hi @jiayihu, we set up a mock `EnvVarQuery` struct that can be used in tests. Here is an example of a test that uses it and of breaking a function into `fn_name` and `inner_fn_name` to create a testable `inner_fn_name` that tests in the mock struct: https://sourcegraph.com/github.com/deislabs/akri/-/blob/agent/src/protocols/mod.rs#L121

Reply:
Thanks for the tip! I have been using `EnvVarQuery` already in some tests, but this case is different. I've shown the problematic line of code in my previous comment, which is not written by me actually. So I would need to pass `EnvVarQuery` down to all the functions until the one actually reading the env variable, since it's an internal method of akri. That would make the API a bit awkward I guess.

I believe there's maybe a better way to write this Akri method, thus why I asked for input from you:
https://github.com/deislabs/akri/blob/1c81b6ca44cccb596f0b19a01a220d96154b7658/agent/src/protocols/mod.rs#L16-L37